### ISSUE TITLE
"Alphanumeric" words include ⍰ characters and characters from non-Latin alphabets

### ISSUE THREAD

**Issue by bwbug** [sts10/tidy#11]
I was doing some tests with the 2021 version of the [Wikipedia-based word frequency lists](https://github.com/IlyaSemenov/wikipedia-word-frequency/tree/master/results).  The trailing frequency values on each row seemed to be interpreted by tidy as part of the word, and apparently the file contained also underscore characters `_` that were recognized as words by tidy, so I tried to deal with these issues using the `-in` flags.  To get a sampling of the results, I ran

`tidy -in --samples --dry-run enwiki-20210820-words-frequency.txt`

On several occasions, the generated passphrases included words using Cyrillic or Greek characters, and even words consisting of characters that could not be represented (displayed using the "squares with question mark" replacement character: ⍰).

Is this expected behavior?  The `-n` flag is supposed to leave behind only "alphanumeric" characters, so `-in` should be limited to characters in the "alphabet".  Did you intend for this to include non-Latin alphabets?  That seems inconsistent with the description of the `-L` flag, which defines "alphabetic" to include only `[A-Z]` and `[a-z]`.  At a minimum, I think it would make sense for `-n` to delete any characters that cannot be represented (⍰).

P.S. Separate issue, but is there a better way to strip the numerical frequency values appended to each row in the Wikipedia data sets? (e.g., something like `-st` but from the back?).

**Comment by sts10** [comment]
Ah, yes, this aspect of Tidy is a bit unfinished. Handling strange characters is a bit tricky in Rust... 

>  The -n flag is supposed to leave behind only "alphanumeric" characters, so -in should be limited to characters in the "alphabet". Did you intend for this to include non-Latin alphabets?

Yes, the `-n` flag should leave characters with diacritics and Cyrillic or Greek characters. I wrote "Characters with diacritics will remain", but I could add more to that sentence if you think it's best. This is an effort to provide functionality for those working with non-English words. I think it's important to have Tidy work for non-English word lists. 

> That seems inconsistent with the description of the -L flag, which defines "alphabetic" to include only [A-Z] and [a-z].

The L flag's help text reads: "Remove all words with any non-alphabetic characters or diacritic characters from new list (leaving only words composed entirely of letters [A-Z] or [a-z])". This is intended to be exactly that much more strict than the `-n` flag, since `-L` also removes diacritic characters. 
 
Maybe i could clean up my use of the term "alphanumeric" in the help text... Admittedly I'm not sure whether to include diacritic and Greek characters in that category. Open to suggestions for more consistent/user-friendly terminology. 

> even words consisting of characters that could not be represented (displayed using the "squares with question mark" replacement character: ⍰).

This feels more like a real bug in Tidy's understanding of characters. Is there some way you could give he a word or line-number in the Wikipedia file that produces this question mark character when run through Tidy? 

> P.S. Separate issue, but is there a better way to strip the numerical frequency values appended to each row in the Wikipedia data sets? (e.g., something like -st but from the back?).

Not that I remember. Like a "delete _after_ space" option? That shouldn't be too hard...

**Comment by sts10** [comment]
I added new options to delete after space, tab and comma: https://github.com/sts10/tidy/commit/5be2a2a9cd8d23af480ee95f8e37575c8492eb51

**Comment by sts10** [comment]
> As far as clarifying the help text for -n, can you explain what specific technical criteria you use to include/exclude characters? That would help me understand what to expect, and would allow me to suggest a more accurate help text.

Here's how they work:

`-n`  only allows characters that return true from [Rust's `is_alphanumeric` method](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphanumeric).

`-L` only allows characters that return true from a function I wrote:
```
pub fn is_alphabetic(chr: u16) -> bool {
    (chr >= 65 && chr <= 90) || (chr >= 97 && chr <= 122)
}
```

the numbers here are [ASCII values (as decimals)](https://ascii-tables.com/). 65 is "A" and 90 is "Z"; 97 is "a" and 122 is "z". I'm remembering now that I wanted a very strict _English_ alphabetic option. This was probably due to frustrations with strange characters that you're describing.

Here's what I propose:  
1. Re-name the strictly [A-Z] and [a-z]  option as `--remove-non-latin-alphabet`. Hopefully this is a more descriptive label (I hope that's the right term?), and we can even preserve the `-L` short name! This will continue to be our most "strict" filter.
2. Write a new `--alphabetic` option that will allow accented characters, namely ASCII values 65 though 90; 97 through 122; and 192 through 255. 
3. Keep the current `-n`/`-N` non-alphanumeric filter as is. 

One bummer even with this new plan is that punctuation (like commas), spaces, and tabs will be caught in all three of these filters (i.e. removed by them). It seems like you're describing use-cases where you'd want to filter out strange characters, but preserve integers, commas, and other standard-ish delimiters like tabs and spaces. Is that right? Any ideas what we should label that option?

**Comment by sts10** [comment]
I also just added a `--remove-non-ascii` option. It uses [Rust's `is_ascii` method](https://doc.rust-lang.org/std/primitive.char.html#method.is_ascii) to  

> Remove all words that have any non-ASCII characters from new list

I think it removes our new strange friend " اج 12 ", but keeps most other characters, including punctuation and whitespace. Interestingly, the Chinese character "京" -- which is alphabetic -- is removed... not sure why. I guess some or all complex Asian characters aren't ASCII.

**Comment by sts10** [comment]
> Was this not the case before, as well? Was there a filter that previously preserved such characters but now no longer would do so using the new plan?

This was the case before, yes. Just noting it for myself mostly. 

> In Issue #8, I try to make a case for splitting the original row data into the "word" (everything up to the first column delimiter) and the "metadata" (everything from the first column delimiter to the end-of-line). Then all filtering (character-level and word-level eliminations or substitutions) would be done only on the "word" part of the input (but obviously, if a "word" is removed from the list, then the corresponding "metadata" must also be removed); after the filtering is done, the "word" and corresponding "metadata" would be joined together again prior to any optional sorting and output of results. If it is possible to implement something like this, then the various filters would not have any adverse effect on the numerical metadata and column delimiters that appear after each "word" (because the filters would act only on the "word" itself).

Ah OK, this is a helpful explanation. While I like the idea, I'm tempted to file this under "outside of Tidy's scope"... I think there are plenty of other tools for sorting comma-separated data, e.g. [qsv](https://github.com/jqnatividad/qsv). Obviously if I could fix #12 this would a more defensible position for me to hold.

> I do not foresee use cases in which tabs, commas, or periods should be preserved when the "word" is filtered. I could envision that some applications would want to preserve space characters (in case expressions like "in vitro" or "ad hoc" are supposed to be valid "words"), but personally I'm no fan of including such constructions in passphrase word lists. Finally, some word lists will probably include contractions (e.g., "don't" or "they'll"); again, I personally wouldn't include such words in a passphrase word list, but you must have wanted to accommodate something like this, since you even included a filter to straighten curly apostrophes.

This sounds right. I'll leave the filters as is for a bit and close this issue for now. Thanks again for your great questions and suggestions!
